eth_utils/abi.py

```diff
@@ -1,68 +1,783 @@
+    Return a list of each ``ABIElement`` that is of type ``abi_type``.
+
+    :param abi_type: Type of ABI element to filter by.
+    :type abi_type: `str`
+    :param contract_abi: Contract ABI.
+    :type contract_abi: `ABI`
+    :return: List of ABI elements of the specified type.
+    :rtype: `Sequence[ABIElement]`
+
+    .. doctest::
+
+        >>> from eth_utils import filter_abi_by_type
+        >>> abi = [
+        ...   {"type": "function", "name": "myFunction", "inputs": [], "outputs": []},
+        ...   {"type": "function", "name": "myFunction2", "inputs": [], "outputs": []},
+        ...   {"type": "event", "name": "MyEvent", "inputs": []}
+        ... ]
+        >>> filter_abi_by_type("function", abi)
+        [{'type': 'function', 'name': 'myFunction', 'inputs': [], 'outputs': []}, \
+{'type': 'function', 'name': 'myFunction2', 'inputs': [], 'outputs': []}]
+    """
+    return [abi for abi in contract_abi if abi["type"] == abi_type]
+
+
+def get_all_function_abis(contract_abi: ABI) -> Sequence[ABIFunction]:
+    """
+    Return interfaces for each function in the contract ABI.
+
+    :param contract_abi: Contract ABI.
+    :type contract_abi: `ABI`
+    :return: List of ABIs for each function interface.
+    :rtype: `Sequence[ABIFunction]`
+
+    .. doctest::
+
+        >>> from eth_utils import get_all_function_abis
+        >>> contract_abi = [
+        ...   {"type": "function", "name": "myFunction", "inputs": [], "outputs": []},
+        ...   {"type": "function", "name": "myFunction2", "inputs": [], "outputs": []},
+        ...   {"type": "event", "name": "MyEvent", "inputs": []}
+        ... ]
+        >>> get_all_function_abis(contract_abi)
+        [{'type': 'function', 'name': 'myFunction', 'inputs': [], 'outputs': []}, \
+{'type': 'function', 'name': 'myFunction2', 'inputs': [], 'outputs': []}]
+    """
+    return [
+        cast(ABIFunction, function_abi)
+        for function_abi in filter_abi_by_type("function", contract_abi)
+    ]
+
+
+def get_all_event_abis(contract_abi: ABI) -> Sequence[ABIEvent]:
+    """
+    Return interfaces for each event in the contract ABI.
+
+    :param contract_abi: Contract ABI.
+    :type contract_abi: `ABI`
+    :return: List of ABIs for each event interface.
+    :rtype: `Sequence[ABIEvent]`
+
+    .. doctest::
+
+        >>> from eth_utils import get_all_event_abis
+        >>> contract_abi = [
+        ...   {"type": "function", "name": "myFunction", "inputs": [], "outputs": []},
+        ...   {"type": "function", "name": "myFunction2", "inputs": [], "outputs": []},
+        ...   {"type": "event", "name": "MyEvent", "inputs": []}
+        ... ]
+        >>> get_all_event_abis(contract_abi)
+        [{'type': 'event', 'name': 'MyEvent', 'inputs': []}]
+    """
+    return [
+        cast(ABIEvent, event) for event in filter_abi_by_type("event", contract_abi)
+    ]
+
+
+def get_normalized_abi_inputs(
+    abi_element: ABIElement,
+    args: Optional[Sequence[Any]] = None,
```

Comment:
I've more commonly seen this style of accept zero or more args/kwargs expressed as:
```suggestion
    *args: Optional[Sequence[Any]],
    **kwargs: Optional[Dict[str, Any]],
```
But I'm curious if you deliberately made the decision to not use that notation? I think using the `*` would save us lines 411 and 412, but not a huge deal either way. I'm mostly just curious :)

Reply:
I thought about that too. It seems more intuitive to me to pass in the *abi element's* args and kwargs as sequence and dict, respectively. If it was for this method itself I think that would make more sense, but it's less intuitive to me to pass the args and kwargs for the abi element that I'm looking for as args and kwargs to the utility function itself. I could be convinced otherwise though.

In this way, it's similar to the `abi.encode()` and `abi.decode()` methods where we pass in the list of args.

Reply:
Sweet, I'm good with that. I like the encode/decode consistency!

Reply:
Although maybe there's something to the `kwargs` being unpacked. Or I can still be convinced I think 😄 

*args*
```py 
# if not defined as *args, **kwargs
get_normalized_abi_inputs(abi, ('myName', 123, ('0x1', 1, b'\x01')))
get_normalized_abi_inputs(abi, {'name': 'myName', 's': 123, 't': ('0x1', 1, b'\x01')}

# defined as *args, **kwargs
get_normalized_abi_inputs(abi, 'myName', 123, ('0x1', 1, b'\x01'))
get_normalized_abi_inputs(abi, name='myName', s=123, t=('0x1', 1, b'\x01'))
```

Thoughts @reedsa @kclowes on what feels better?

Reply:
The args and kwargs were already defined this way, so no thought crossed my mind there. I see some inconsistency in usage from web3 as well.

In the `_encode_data_in_transaction` function (https://github.com/ethereum/web3.py/blob/39cde3b8894d71a6663dd1588a95dc5a62197d96/web3/contract/base_contract.py#L1081), the original `merge_args_and_kwargs` is called. The arguments are already defined with `_encode_data_in_transaction(self, *args: Any, **kwargs: Any)`.

Another case doesnt use that style unfortunately: https://github.com/ethereum/web3.py/blob/39cde3b8894d71a6663dd1588a95dc5a62197d96/web3/contract/base_contract.py#L898

It'd be nice to clean those up in the end. I think these particular cases args and kwargs should consistenly use `*args` and `**kwargs`. Let me know if that sounds good!

Reply:
@reedsa color me convinced honestly. Feels more pythonic and just a nicer approach in general. If anyone has the argset as a sequence, it can still be unpacked pretty seamlessly: 

```py
get_normalized_abi_inputs(abi, *('myName', 123, ('0x1', 1, b'\x01')))
```

Reply:
> I think these particular cases args and kwargs should consistenly use *args and **kwargs. Let me know if that sounds good!

That sounds good to me! I agree that it feels more pythonic to use the `*` syntax, and whatever we land on should be consistent throughout 👍

Reply:
I'm also fine if we want to standardize this in a separate PR. If it makes more sense to do it here, that's fine too.

Reply:
I've gone through the `*args` and `**kwargs` changes and checked it all works from the `web3` side. 🥳